pkg/dependencies/dependents.go

```diff
@@ -0,0 +1,237 @@
+//
+// Returns:
+// - A slice of PackageName, sorted by DependentCount in descending order.
+// - An error
+func GetDependenciesBySortedDependentCnt(ctx context.Context, gqlClient graphql.Client) ([]PackageName, error) {
+	packages, err := findDependentsOfDependencies(ctx, gqlClient)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get dependents: %v", err)
+	}
+
+	var packagesArr []PackageName
+
+	for n, d := range packages {
+		packagesArr = append(packagesArr, PackageName{Name: n, DependentCount: len(d.dependents)})
+	}
+
+	sort.Slice(packagesArr, func(i, j int) bool {
+		return packagesArr[i].DependentCount > packagesArr[j].DependentCount
+	})
+	return packagesArr, nil
+}
+
+// findDependentsOfDependencies queries the GraphQL endpoint to retrieve all SBOMs and constructs
+// a map of dependencyNode, where each node represents a package and its dependents. This function is designed to
+// identify and map out the relationships between packages based on the SBOM data. It filters out inconsistent data
+// from "deps.dev" and handles packages with and without specific version IDs. The function employs a breadth-first search
+// (BFS) algorithm to traverse the dependency graph and populate the map with packages and their respective dependents.
+//
+// Returns:
+//   - A map where the key is a string representing the package name, and the value is a dependencyNode containing
+//     the dependents of the package.
+//   - An error
+func findDependentsOfDependencies(ctx context.Context, gqlClient graphql.Client) (map[string]dependencyNode, error) {
+	// Initialize a map to hold package names mapped to their dependency nodes. This will be the returned map
+	packages := make(map[string]dependencyNode)
+
+	// Initialize maps to hold dependency and dependent edges.
+	// dependentEdges maps a package ID to the IDs of packages that depend on it (DepPkg to Pkg).
+	// dependencyEdges maps a package ID to the IDs of its dependencies (Pkg to DepPkg).
+	dependentEdges := make(map[string][]string)  // id -> dependents ids
+	dependencyEdges := make(map[string][]string) // id -> dependency ids
+
+	// Initialize a map to convert package IDs to their names.
+	idToName := make(map[string]string)
+
+	sboms, err := model.HasSBOMs(ctx, gqlClient, model.HasSBOMSpec{})
+
+	if err != nil {
+		return nil, fmt.Errorf("error getting dependencies: %v", err)
+	}
+
+	for _, resp := range sboms.HasSBOM {
+		// Skip entries from "deps.dev" because they are inconsistent.
+		if resp.Origin == deps_dev.DepsCollector {
+			continue
+		}
+		// Iterate through the included dependencies of each SBOM.
+		for _, dependency := range resp.IncludedDependencies {
+			// Construct unique names for the dependency package and the package itself.
+			// TODO: Make the names actually unique, not just add "_".
+			depPkgName := dependency.DependencyPackage.Type + "_" + dependency.DependencyPackage.Namespaces[0].Namespace + "_" + dependency.DependencyPackage.Namespaces[0].Names[0].Name
```

Comment:
Is it possible to have the API endpoint output the package as a purl?

Reply:
I don’t think that we can output as a purl because we have two types of data for each package, the package name, and the number of dependents.

Reply:
+1 on outputting the package name as a PURL instead of `_` delimetered strings (see https://github.com/guacsec/guac/blob/main/pkg/assembler/helpers/purl.go#L47). I'm not sure what you mean we can't do this because we have two dimensions of data? Would it be a PURL to data mapping?

Reply:
Oh, I don't think I understood what @mdeicas meant before; I have updated the spec so that it outputs the packages as purls. Thanks!